src/lib.rs

```diff
@@ -465,6 +471,12 @@ impl<D: tb::Bool, H: tb::Bool, T: tb::Bool> InvoiceBuilder<D, H, T> {
 		self
 	}
 
+	/// Sets the payment secret
+	pub fn payment_secret(mut self, payment_secret: PaymentSecret) -> Self {
```

Comment:
Shouldn't this get a new type boolean to enforce that only, at max, one payment_secret is added. Same goes for `payee_pub_key`, actually.

Reply:
Hm, I'm not that sure anymore if this API was a wise choice. When I used it with conditional parameters it was needlessly cumbnersome and it gets worse the more parameters there are (since a cop-out is setting the conditional stuff at the end, but that only works once). 

I think it's even worse for " 0 or 1" than for "1 of these" cases:

```rust
let builder: Builder<False> = Builder::new();

let builder = if condition {
    builder.set_with_a(…) // Builder<True>
} else{
    builder.set_with_b(…) // Builder<True>
};

builder.build()
```

Here the types can work out, since the parameter has to be flipped to true at some point anyway and it can be done in the same place.

```rust
let builder: Builder<False> = Builder::new();

let builder = if condition {
    builder.set_with_a(…) // Builder<True>
} else {
    builder // Builder<False>
}; // ERROR: mismatched types

builder.build()
```

If a parameter is truly optional there's suddenly a divergence of types till `build()` is called, so if there are multiple of these decisions these can't be stacked linearly, but have to be nested so that `build()` can be called inside each branch to get to a unified type again.

It's a mess. I hope that this was the reason for me not to impl type booleans for payee pub key back then and not that I simply overlooked it :smile: but yeah, I guess we'll need to catch this in `build()`, adding a variant to `CreationError`.

Reply:
Ah, right, that does make it impractical. Indeed, probably points to the need for some other way to enforce this. Maybe instead of storing the data for payment pubkey and secret in the Vec we can put them in their own `Option`s in the struct and then just let the `set` methods set-or-replace.

Reply:
@TheBlueMatt: @sgeisler confirmed he's fine to punt the changes in this discussion to follow-up. Lmk if you're fine with that too.

Reply:
No, probably should be resolved in a separate PR, though its kinda sad you can build bogus invoices right now.